File: standard-udfs-hive/src/test/java/com/linkedin/stdudfs/hive/examples/TestStructCreateByNameFunctionWrapper.java
package com.linkedin.stdudfs.hive.examples;

import com.google.common.collect.ImmutableList;
import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
import org.testng.annotations.Test;

import static com.linkedin.stdudfs.hive.common.AssertHiveGenericUdf.*;
import static com.linkedin.stdudfs.hive.common.StandardObjectInspectorFactory.*;


public class TestStructCreateByNameFunctionWrapper {
  @Test
  public void testStructCreateByName() {
    assertFunction(
        new StructCreateByNameFunctionWrapper(),
        new ObjectInspector[]{
            STRING,
            STRING,
            STRING,
            STRING
        },
        new Object[]{"a", "x", "b", "y"},
        ImmutableList.of("x", "y")
    );

    assertFunction(
        new StructCreateByNameFunctionWrapper(),
        new ObjectInspector[]{
            STRING,
            STRING,
            STRING,
            STRING
        },
        new Object[]{null, "x", "b", "y"},
        null
    );

    assertFunction(
        new StructCreateByNameFunctionWrapper(),
        new ObjectInspector[]{
            STRING,
            STRING,
            STRING,
            STRING
        },
        new Object[]{"a", null, "b", "y"},
        null
    );

    assertFunction(
        new StructCreateByNameFunctionWrapper(),
        new ObjectInspector[]{
            STRING,
            STRING,
            STRING,
            STRING
        },
        new Object[]{"a", "x", null, "y"},
        null
    );

    assertFunction(
        new StructCreateByNameFunctionWrapper(),
        new ObjectInspector[]{
            STRING,
            STRING,
            STRING,
            STRING
        },
        new Object[]{"a", "x", "b", null},
        null
    );
  }
}

File: standard-udfs-spark/src/test/scala/com/linkedin/stdudfs/spark/examples/TestStructCreateByNameFunctionWrapper.scala
package com.linkedin.stdudfs.spark.examples

import com.linkedin.stdudfs.spark.common.AssertSparkExpression._
import org.testng.annotations.{BeforeClass, Test}

class TestStructCreateByNameFunctionWrapper {

  @BeforeClass
  def registerFunction(): Unit = {
    registerStandardUdf(
      "struct_create_by_name",
      classOf[StructCreateByNameFunctionWrapper]
    )
  }

  @Test
  def testStructCreateByNameFunction(): Unit = {
    assertFunction("cast(struct_create_by_name('a', 'x', 'b', 'y') as struct<a:string, b:string>).a", "x")
    assertFunction("cast(struct_create_by_name('a', 'x', 'b', 'y') as struct<a:string, b:string>).b", "y")
    assertFunction("struct_create_by_name(null, 'x', 'b', 'y')", null)
    assertFunction("struct_create_by_name('a', 'x', null, 'y')", null)
  }
}
